shelley/chain-and-ledger/formal-spec/multi-sig.tex

```diff
@@ -0,0 +1,567 @@
+  bookmarks=true,
+  colorlinks=false,
+  linkcolor={blue},
+  citecolor={blue},
+  urlcolor={blue},
+  linkbordercolor={white},
+  citebordercolor={white},
+  urlbordercolor={white}
+}
+
+\title{A Formal Specification of a Multi-Signature Scheme using Scripts}
+
+\author{Jared Corduan  \\ {\small \texttt{jared.corduan@iohk.io}} \\
+   \and Matthias G\"udemann  \\ {\small \texttt{matthias.gudemann@iohk.io}}}
+
+%\date{}
+
+\maketitle
+
+\begin{abstract}
+  This documents specifies a simple way to support multi-signature
+  transactions. It is based on a simple script model which uses only single step
+  script execution and does not require data scripts.
+
+  A multi-signature scheme allows an unspent transaction output to be used as an
+  input to a new transaction if a pre-defined combination of signatures is
+  provided, e.g., two persons have to sign simultaneously, two out of three
+  keys have to be provided, etc.
+
+  For an output that is locked with a multi-signature, the set of all keys that
+  signed the transaction is given to the script for validation. In this way, the
+  script can decide in a single step whether is has the correct key combination
+  to spend the output.
+\end{abstract}
+
+% \section*{List of Contributors}
+% \label{acknowledgements}
+
+% \tableofcontents
+% \listoffigures
+
+\section{Introduction}
+\label{sec:introduction}
+
+This specification for a simple multi-signature scheme is based
+on~\cite{shelley_formal_spec} which formally specifies the Shelley Cardano
+ledger. The main changes are the following:
+
+\begin{itemize}
+\item Add a new address type that for outputs locked by scripts.
+\item Change the type of transaction inputs of a transaction to optionally
+  include a validator script.
+\item Adapt the transaction validation in such a way that funds locked by a
+  multi-signature script can be spent.
+\item Adapt the functions used in the validation for the extended types of
+  addresses and transaction inputs with scripts.
+\end{itemize}
+
+In this approach for multi-signature, the scripts receive a set of hashed keys
+where each corresponding key was used to correctly sign the transaction. The
+script can then simply check the hashes against its own representation of which
+keys in which combination can unlock the unspent output.
+
+This means that the scripts are completely stateless and no data needs to be
+supplied to the script apart from the hashed keys used to sign the transaction.
+It allows for any type of $n$ out of $m$ required signatures for a specific
+unspent transaction output.
+
+\section{Types}
+\label{sec:types}
+
+\begin{figure*}[hbt]
+  \emph{Abstract types}
+
+  \begin{equation*}
+    \begin{array}{r@{~\in~}lr}
+      script & \Script & \text{script type}
+    \end{array}
+  \end{equation*}
+
+  \emph{Derived types}
+
+  \begin{equation*}
+    \begin{array}{r@{~\in~}l@{\qquad=\qquad}lr}
+      \var{addr} & \AddrScr & \HashScr \times \Script & \text{Script address} \\
```

Comment:
I do not think we need to include the script in the address. It can be revealed on use, since it's now attached to a txin.